File: backend/src/main/java/com/NBE_4_5_2/Team5/global/security/CorsConfig.java
package com.NBE_4_5_2.Team5.global.security;

import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;

import jakarta.servlet.http.HttpServletRequest;
import lombok.NonNull;

@Configuration
public class CorsConfig implements CorsConfigurationSource {
	@Value("${custom.front.host}")
	private String frontHost;

	private String ALLOWED_ORIGIN;
	private static final List<String> ALLOWED_METHODS = List.of("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS");

	@Override
	public CorsConfiguration getCorsConfiguration(@NonNull HttpServletRequest request) {
		CorsConfiguration config = new CorsConfiguration();
		ALLOWED_ORIGIN = "https://%s".formatted(frontHost);
		config.setAllowedOrigins(List.of(ALLOWED_ORIGIN, "http://localhost:3000"));
		config.setAllowedMethods(ALLOWED_METHODS);
		config.setAllowCredentials(true);
		config.setAllowedHeaders(List.of("Authorization", "Content-Type", "Accept", "Origin"));
		config.setExposedHeaders(List.of("Authorization"));
		config.setMaxAge(3600L);

		return config;
	}
}

File: backend/src/main/java/com/NBE_4_5_2/Team5/global/security/CorsConfig.kt
package com.NBE_4_5_2.Team5.global.security

import jakarta.servlet.http.HttpServletRequest
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import java.util.List

@Configuration
class CorsConfig(
    @Value("\${custom.front.host}")
    private val frontHost: String
) : CorsConfigurationSource {

    companion object {
        private val ALLOWED_METHODS    = listOf("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
        private val ALLOWED_HEADERS    = listOf("Authorization", "Content-Type", "Accept", "Origin")
        private val EXPOSED_HEADERS    = listOf("Authorization")
        private const val MAX_AGE: Long = 3600L
    }

    override fun getCorsConfiguration(request: HttpServletRequest): CorsConfiguration =
        CorsConfiguration().apply {
            // 매번 호출 시점에 frontHost 로부터 allowedOrigin 계산
            val allowedOrigin = "https://$frontHost"
            allowedOrigins = listOf(allowedOrigin, "http://localhost:3000")
            allowedMethods = ALLOWED_METHODS
            allowCredentials = true
            allowedHeaders = ALLOWED_HEADERS
            exposedHeaders = EXPOSED_HEADERS
            maxAge = MAX_AGE
        }
}
